dbms/src/Flash/Pipeline/Schedule/WaitQueue.cpp

```diff
@@ -0,0 +1,69 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Flash/Pipeline/Schedule/WaitQueue.h>
+#include <assert.h>
+#include <common/likely.h>
+
+namespace DB
+{
+bool WaitQueue::take(std::list<TaskPtr> & local_waiting_tasks)
+{
+    {
+        std::unique_lock lock(mu);
+        while (true)
+        {
+            if (unlikely(is_closed))
+                return false;
+            if (!waiting_tasks.empty() || !local_waiting_tasks.empty())
```

Comment:
How can `local_waiting_tasks` not empty at this point?

Reply:
The task that returned `STATUS::WAITING` from await in the previous round of loop may still be kept in `local_waiting_tasks` .